File: fixtures/scale/swift/medium/Eval/Evaluator.swift
final class Evaluator {
    let environment: Environment
    let functions: FunctionRegistry

    init(environment: Environment, functions: FunctionRegistry) {
        self.environment = environment
        self.functions = functions
    }

    func eval(_ node: Node) throws -> Double {
        switch node {
        case let n as NumberLiteral: return n.value
        case let n as VariableRef: return try environment.resolve(n.name)
        case let n as UnaryOp: return -(try eval(n.operand))
        case let n as BinaryOp: return try BinaryMath.apply(n.op, try eval(n.left), try eval(n.right))
        case let n as FunctionCall:
            let arguments = try n.arguments.map { try eval($0) }
            return try functions.invoke(n.name, arguments)
        case let n as Assignment:
            let value = try eval(n.value)
            environment.assign(n.name, value)
            return value
        default:
            throw EvalError(message: "unsupported node", subject: node.describe())
        }
    }
}

File: fixtures/scale/php/medium/src/Eval/Evaluator.php
<?php

namespace Calc\Eval;

use Calc\Ast\Assignment;
use Calc\Ast\BinaryOp;
use Calc\Ast\FunctionCall;
use Calc\Ast\Node;
use Calc\Ast\NumberLiteral;
use Calc\Ast\UnaryOp;
use Calc\Ast\VariableRef;
use Calc\Functions\FunctionRegistry;

require_once __DIR__ . '/BinaryMath.php';

class Evaluator
{
    public function __construct(private Environment $environment, private FunctionRegistry $functions) {}

    public function eval(Node $node): float
    {
        $args = fn (FunctionCall $n) => array_map(fn (Node $a) => $this->eval($a), $n->arguments);
        return match (true) {
            $node instanceof NumberLiteral => $node->value,
            $node instanceof VariableRef => $this->environment->resolve($node->name),
            $node instanceof UnaryOp => -$this->eval($node->operand),
            $node instanceof BinaryOp => BinaryMath::apply($node->op, $this->eval($node->left), $this->eval($node->right)),
            $node instanceof FunctionCall => $this->functions->invoke($node->name, $args($node)),
            $node instanceof Assignment => $this->assign($node),
            default => throw new EvalError('unsupported node', $node->describe()),
        };
    }

    private function assign(Assignment $node): float
    {
        $value = $this->eval($node->value);
        $this->environment->assign($node->name, $value);
        return $value;
    }
}
